File: compiler/src/queries/BoundQueries.swift
extension Queries {
    public static var resolvedBound: Query<Bound, ResolvedBound> {
        { db, node, body in
            guard let bounds = db[node, Bounds.self]?.bounds else { return }

            for case (_, (let bound, let resolved?)) in bounds { body(bound, resolved) }
        }
    }

    public static var unresolvedBound: Query<Bound> {
        { db, node, body in
            guard let bounds = db[node, Bounds.self]?.bounds else { return }

            for case (_, (let bound, nil)) in bounds { body(bound) }
        }
    }
}

File: compiler/queries/src/bound_queries.rs
use crate::QueryCtx;
use wipple_core::{
    db::Node,
    typecheck::bounds::{ResolvedBound, ResolvedBounds, UnresolvedBound},
};

pub fn resolved_bounds<'a>(db: &QueryCtx<'a>, node: Node) -> Vec<&'a ResolvedBound> {
    let Some(ResolvedBounds(bounds)) = db.get(node) else {
        return Vec::new();
    };

    bounds
        .values()
        .filter_map(|result| result.as_ref().ok())
        .collect()
}

pub fn unresolved_bounds<'a>(db: &QueryCtx<'a>, node: Node) -> Vec<&'a UnresolvedBound> {
    let Some(ResolvedBounds(bounds)) = db.get(node) else {
        return Vec::new();
    };

    bounds
        .values()
        .filter_map(|result| result.as_ref().err())
        .collect()
}
